File: java_source/lectures/src/main/java/union_find/QuickUnion.java
package union_find;

public class QuickUnion {
    private int[] id;

    public QuickUnion(int N) {
        this.id = new int[N];

        for (int i = 0; i < N; i++) {
            this.id[i] = i;
        }
    }

    public boolean connected(int p, int q) {
      return this.root(p) == this.root(q);
    }

    public void union(int p, int q) {
      int rootP = this.root(p);
      int rootQ = this.root(q);

      this.id[rootP] = rootQ;
    }

    private int root(int i) {
      while (i != this.id[i]) {
        i = this.id[i];
      }

      return i;
    }
  }

File: ts/src/union_find/quick_union.ts
export default class QuickUnionUF {
  private id: Array<number> = [];

  constructor(N: number) {
    for (let i: number = 0; i < N; i++) {
      this.id.push(i);
    }
  }

  public connected(p: number, q: number): boolean {
    return this.root(p) === this.root(q);
  }

  public union(p: number, q: number): void {
    const rootP = this.root(p);
    const rootQ = this.root(q);

    this.id[rootP] = rootQ;
  }

  private root(i: number): number {
    while (i != this.id[i]) {
      i = this.id[i];
    }

    return i;
  }
}
